server/src/main/java/org/elasticsearch/gateway/GatewayService.java

```diff
@@ -256,6 +256,9 @@ public ClusterState execute(final ClusterState currentState) {
         @Override
         public void clusterStateProcessed(final String source, final ClusterState oldState, final ClusterState newState) {
             logger.info("recovered [{}] indices into cluster_state", newState.metaData().indices().size());
+            // reset flag even though state recovery completed, to ensure that if we subsequently become leader again based on a
+            // not-recovered state, that we again do another state recovery.
+            resetRecoveredFlags();
```

Comment:
Would it make sense to rename the `recovered` field to `recoveryAllowed` (or something along those lines) since it's not really a one time thing to run recovery now?